File: evidenciar-pdf/scripts/gerar_pdf.js
// Gera um PDF de evidências com a identidade da PersonalizeIT a partir de um evidencia.json.
//
// Uso:  node gerar_pdf.js <evidencia.json> [pastaDasImagens]
//
// Renderiza HTML e imprime com o Chrome (Playwright). Não depende de Word/LibreOffice.
// A marca (cores, fonte, logo) vem de ../assets/brand.json — edite lá, não aqui.
//
// evidencia.json (só title e output são obrigatórios):
// {
//   "title": "Permissões do Relatório de Tickets",
//   "subtitle": "Evidências da entrega",         // opcional, aceita \n
//   "client": "Supermercado Brunetto",           // aparece na capa e no rodapé
//   "tags": ["PER-102", "PER-103"],              // chips na capa (IDs das tarefas)
//   "dateline": "Setembro de 2026",              // default: mês/ano de hoje
//   "intro": { "heading": "...", "paragraphs": [...], "bullets": [...] },
//   "summary": [["Item", "Situação"], ["...", "..."]],   // 1ª linha = cabeçalho
//   "sections": [{
//      "heading": "1. ...",
//      "level": 1,                                // 1 = nova página (default), 2 = segue na mesma
//      "paragraphs": ["..."],
//      "bullets": ["..."],
//      "steps": ["..."],
//      "note": { "type": "info|ok|alerta", "title": "...", "text": "..." },
//      "figures": [{ "img": "01.png", "caption": "Figura 1 — ..." }]
//   }],
//   "glossary": [["Termo", "definição"]],
//   "output": "D:/.../Evidencias.pdf"
// }
const fs = require('fs');
const path = require('path');
const { chromium } = require('playwright');

const ROOT = path.resolve(__dirname, '..');
const brand = JSON.parse(fs.readFileSync(path.join(ROOT, 'assets', 'brand.json'), 'utf8'));

const specFile = process.argv[2];
if (!specFile) {
  console.error('uso: node gerar_pdf.js <evidencia.json> [pastaDasImagens]');
  process.exit(1);
}
const spec = JSON.parse(fs.readFileSync(specFile, 'utf8'));
const IMGDIR = process.argv[3] || path.dirname(path.resolve(specFile));

const esc = (s) =>
  String(s ?? '').replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));

// **negrito** e `código` no texto dos parágrafos/bullets
const rich = (s) =>
  esc(s)
    .replace(/\*\*(.+?)\*\*/g, '<strong>$1</strong>')
    .replace(/`(.+?)`/g, '<code>$1</code>');

function dataUri(file) {
  const abs = path.isAbsolute(file) ? file : path.join(IMGDIR, file);
  if (!fs.existsSync(abs)) {
    console.warn('  ! imagem não encontrada:', abs);
    return null;
  }
  const ext = path.extname(abs).slice(1).toLowerCase();
  const mime = ext === 'svg' ? 'image/svg+xml' : ext === 'jpg' ? 'image/jpeg' : `image/${ext}`;
  return `data:${mime};base64,${fs.readFileSync(abs).toString('base64')}`;
}

const logo = dataUri(path.join(ROOT, 'assets', brand.logo));

// A fonte da marca é comercial e não é versionada: se os .ttf existirem nesta máquina,
// são embutidos; se não, a stack do brand.json cai no fallback sem quebrar o layout.
function fontFaces() {
  const files = brand.fontFiles || {};
  const nome = (brand.font.match(/"([^"]+)"/) || [null, 'Marca'])[1];
  const faces = Object.entries(files)
    .map(([peso, arq]) => {
      if (!fs.existsSync(arq)) return null;
      const b64 = fs.readFileSync(arq).toString('base64');
      return `@font-face{font-family:"${nome}";font-weight:${peso};font-style:normal;
        src:url(data:font/ttf;base64,${b64}) format("truetype");}`;
    })
    .filter(Boolean);
  if (!faces.length) console.warn(`  ! fonte "${nome}" não encontrada nesta máquina — usando o fallback da stack`);
  return faces.join('\n');
}

function figuras(list = []) {
  return list
    .map((f) => {
      const src = dataUri(f.img);
      if (!src) return '';
      return `<figure class="fig">
        <img src="${src}" alt="">
        ${f.caption ? `<figcaption>${rich(f.caption)}</figcaption>` : ''}
      </figure>`;
    })
    .join('\n');
}

function nota(n) {
  if (!n) return '';
  const tipo = n.type || 'info';
  return `<aside class="nota nota--${esc(tipo)}">
    ${n.title ? `<p class="nota__t">${rich(n.title)}</p>` : ''}
    <p class="nota__x">${rich(n.text || '')}</p>
  </aside>`;
}

function tabela(rows = []) {
  if (!rows.length) return '';
  const [head, ...body] = rows;
  return `<table class="tab">
    <thead><tr>${head.map((c) => `<th>${rich(c)}</th>`).join('')}</tr></thead>
    <tbody>${body
      .map((r) => `<tr>${r.map((c) => `<td>${rich(c)}</td>`).join('')}</tr>`)
      .join('')}</tbody>
  </table>`;
}

function bloco(s) {
  return [
    (s.paragraphs || []).map((p) => `<p>${rich(p)}</p>`).join(''),
    (s.bullets || []).length ? `<ul>${s.bullets.map((b) => `<li>${rich(b)}</li>`).join('')}</ul>` : '',
    (s.steps || []).length ? `<ol>${s.steps.map((b) => `<li>${rich(b)}</li>`).join('')}</ol>` : '',
    s.table ? tabela(s.table) : '',
    nota(s.note),
    figuras(s.figures),
  ].join('\n');
}

const hoje = new Date();
const MESES = ['janeiro', 'fevereiro', 'março', 'abril', 'maio', 'junho',
  'julho', 'agosto', 'setembro', 'outubro', 'novembro', 'dezembro'];
const dateline = spec.dateline || `${MESES[hoje.getMonth()]} de ${hoje.getFullYear()}`;

const secoes = (spec.sections || [])
  .map((s, i) => {
    const lvl = s.level || 1;
    const quebra = lvl === 1 && i > 0 ? ' quebra' : '';
    return `<section class="sec${quebra}">
      <h${lvl === 1 ? 2 : 3}>${rich(s.heading || '')}</h${lvl === 1 ? 2 : 3}>
      ${bloco(s)}
    </section>`;
  })
  .join('\n');

const html = `<!doctype html>
<html lang="pt-BR"><head><meta charset="utf-8">
<title>${esc(spec.title)}</title>
<style>
  ${fontFaces()}

  @page { size: A4; margin: 18mm 16mm 20mm; }
  @page :first { margin: 0; }

  :root {
    --primaria: ${brand.colors.primary};
    --primaria-escura: ${brand.colors.primaryDark};
    --secundaria: ${brand.colors.secondary};
    --texto: ${brand.colors.text};
    --suave: ${brand.colors.muted};
    --linha: ${brand.colors.line};
    --fundo: ${brand.colors.surface};
    --capa-fundo: ${brand.cover.bg};
    --capa-fundo-2: ${brand.cover.bgSoft};
    --capa-texto: ${brand.cover.text};
    --capa-acento: ${brand.cover.accent};
  }

  * { box-sizing: border-box; }
  body {
    margin: 0; color: var(--texto);
    font-family: ${brand.font};
    font-size: 10.5pt; line-height: 1.55;
    -webkit-print-color-adjust: exact; print-color-adjust: exact;
  }

  /* ---------- capa (grafite: a base da marca; laranja só como acento) ---------- */
  .capa {
    position: relative; overflow: hidden;
    height: 297mm; padding: 26mm 24mm 22mm;
    display: flex; flex-direction: column;
    background: var(--capa-fundo);
    color: var(--capa-texto); page-break-after: always;
  }
  /* faixa de acento no topo — o laranja entra aqui, em área mínima */
  .capa::before {
    content: ""; position: absolute; top: 0; left: 0; right: 0; height: 3mm;
    background: var(--capa-acento);
  }
  /* camada de profundidade em grafite-superfície, canto inferior direito */
  .capa::after {
    content: ""; position: absolute; right: -40mm; bottom: -60mm;
    width: 150mm; height: 150mm; border-radius: 50%;
    background: var(--capa-fundo-2);
  }
  .capa > * { position: relative; z-index: 1; }

  .capa__logo { height: 9mm; width: auto; object-fit: contain; align-self: flex-start; }
  .capa__meio { margin-top: auto; }
  .capa__eyebrow {
    font-size: 9pt; letter-spacing: .2em; text-transform: uppercase;
    color: var(--capa-acento); font-weight: 600; margin: 0 0 6mm;
  }
  .capa h1 {
    font-size: 29pt; line-height: 1.18; font-weight: 700;
    margin: 0 0 6mm; max-width: 21ch;
  }
  .capa__regua { width: 24mm; height: 1.2mm; background: var(--capa-acento); margin: 7mm 0; }
  .capa__sub {
    font-size: 12.5pt; margin: 0; white-space: pre-line;
    color: rgba(255,255,255,.82); line-height: 1.5;
  }
  .chips { display: flex; flex-wrap: wrap; gap: 3mm; margin: 0 0 9mm; padding: 0; list-style: none; }
  .chips li {
    border: 1px solid var(--capa-acento); color: var(--capa-acento);
    border-radius: 999px; padding: 1.3mm 4mm; font-size: 8.5pt;
    letter-spacing: .06em; font-weight: 600;
  }
  .capa__pe {
    margin-top: auto; display: flex; justify-content: space-between; align-items: flex-end;
    font-size: 9.5pt; color: rgba(255,255,255,.7);
    border-top: 1px solid rgba(255,255,255,.18); padding-top: 5mm;
  }

  /* ---------- conteúdo ---------- */
  /* título em grafite com régua laranja: o acento entra na régua, não no texto —
     é o que mantém o laranja dentro dos ~10% da composição que o brand book pede */
  h2 {
    font-size: 15pt; color: var(--secundaria); font-weight: 700; margin: 0 0 4.5mm;
    padding-bottom: 2.5mm; border-bottom: 2px solid var(--primaria);
  }
  h3 { font-size: 11.5pt; color: var(--primaria-escura); margin: 7mm 0 3mm; }
  p { margin: 0 0 3.5mm; }
  ul, ol { margin: 0 0 4mm; padding-left: 6mm; }
  li { margin-bottom: 2mm; }
  li::marker { color: var(--primaria); font-weight: 600; }
  strong { color: var(--primaria-escura); }
  code {
    font-family: Consolas, "Courier New", monospace; font-size: 9pt;
    background: var(--fundo); border: 1px solid var(--linha);
    border-radius: 3px; padding: .3mm 1.4mm;
  }
  .sec { margin-bottom: 8mm; }
  .quebra { page-break-before: always; }

  .intro {
    background: var(--fundo); border-left: 3px solid var(--primaria);
    border-radius: 0 3mm 3mm 0; padding: 6mm 7mm; margin-bottom: 8mm;
  }
  .intro h2 { border: 0; padding: 0; margin-bottom: 3mm; color: var(--primaria-escura); }

  .fig {
    margin: 5mm 0 7mm; page-break-inside: avoid; text-align: center;
  }
  .fig img {
    max-width: 100%; border: 1px solid var(--linha); border-radius: 2.5mm;
    box-shadow: 0 1mm 3mm rgba(0,0,0,.09);
  }
  figcaption {
    font-size: 8.5pt; color: var(--suave); margin-top: 2.5mm;
    font-style: italic; text-align: center;
  }

  .tab { width: 100%; border-collapse: collapse; margin: 0 0 5mm; font-size: 9.5pt; }
  .tab th {
    background: var(--primaria); color: #fff; text-align: left;
    padding: 2.5mm 3mm; font-weight: 600;
  }
  .tab td { padding: 2.5mm 3mm; border-bottom: 1px solid var(--linha); vertical-align: top; }
  .tab tr:nth-child(even) td { background: var(--fundo); }

  .nota {
    border-radius: 2.5mm; padding: 4mm 5mm; margin: 4mm 0 5mm;
    page-break-inside: avoid; border-left: 3px solid var(--primaria);
    background: var(--fundo);
  }
  .nota--ok { border-left-color: ${brand.colors.ok}; }
  .nota--alerta { border-left-color: ${brand.colors.warn}; }
  .nota__t { font-weight: 600; color: var(--primaria-escura); margin: 0 0 1.5mm; }
  .nota--ok .nota__t { color: ${brand.colors.ok}; }
  .nota--alerta .nota__t { color: ${brand.colors.warn}; }
  .nota__x { margin: 0; }

  .gloss dt { font-weight: 600; color: var(--primaria-escura); margin-top: 3mm; }
  .gloss dd { margin: 0 0 2mm; }
</style></head>
<body>

<div class="capa">
  ${logo ? `<img class="capa__logo" src="${logo}" alt="">` : ''}
  <div class="capa__meio">
    <p class="capa__eyebrow">Relatório de evidências</p>
    <h1>${esc(spec.title)}</h1>
    <div class="capa__regua"></div>
    ${spec.subtitle ? `<p class="capa__sub">${esc(spec.subtitle)}</p>` : ''}
  </div>
  <div style="margin-top:auto">
    ${(spec.tags || []).length ? `<ul class="chips">${spec.tags.map((t) => `<li>${esc(t)}</li>`).join('')}</ul>` : ''}
    <div class="capa__pe">
      <span>${esc(spec.client || brand.company)}</span>
      <span>${esc(dateline)}</span>
    </div>
  </div>
</div>

${spec.intro ? `<div class="intro">
  <h2>${rich(spec.intro.heading || 'Resumo')}</h2>
  ${bloco(spec.intro)}
</div>` : ''}

${spec.summary ? `<section class="sec"><h2>Resumo da entrega</h2>${tabela(spec.summary)}</section>` : ''}

${secoes}

${(spec.glossary || []).length ? `<section class="sec quebra">
  <h2>Glossário</h2>
  <dl class="gloss">${spec.glossary.map(([t, d]) => `<dt>${rich(t)}</dt><dd>${rich(d)}</dd>`).join('')}</dl>
</section>` : ''}

</body></html>`;

(async () => {
  const out = path.resolve(spec.output);
  fs.mkdirSync(path.dirname(out), { recursive: true });

  const debugHtml = process.env.KEEP_HTML && path.join(path.dirname(out), path.basename(out, '.pdf') + '.html');
  if (debugHtml) fs.writeFileSync(debugHtml, html, 'utf8');

  const browser = await chromium.launch({ channel: process.env.BROWSER_CHANNEL || 'chrome', headless: true });
  const page = await browser.newPage();
  await page.setContent(html, { waitUntil: 'networkidle' });

  const rodape = `
    <div style="width:100%;font-size:7.5pt;color:${brand.colors.muted};
                font-family:${brand.font.replace(/"/g, "'")};
                padding:0 16mm;display:flex;justify-content:space-between;align-items:center;">
      <span>${esc(spec.title)}</span>
      <span>${esc(spec.client || brand.company)} &nbsp;&ndash;&nbsp; <span class="pageNumber"></span>/<span class="totalPages"></span></span>
    </div>`;

  await page.pdf({
    path: out,
    format: 'A4',
    printBackground: true,
    displayHeaderFooter: true,
    headerTemplate: '<div></div>',
    footerTemplate: rodape,
    margin: { top: '18mm', bottom: '20mm', left: '0mm', right: '0mm' },
  });

  await browser.close();
  const kb = (fs.statSync(out).size / 1024).toFixed(0);
  const nFigs = (spec.sections || []).reduce((n, s) => n + (s.figures || []).length, 0);
  console.log(`PDF salvo em: ${out}  (${kb} KB, ${nFigs} figuras)`);
  if (debugHtml) console.log(`HTML de depuração: ${debugHtml}`);
})();

File: acompanhar-projeto/scripts/gerar_pdf.py
r"""
Monta um documento de acompanhamento em PDF a partir de um documento.json.

    <python> gerar_pdf.py documento.json [pasta-das-imagens]

A marca do documento é o PRODUTO, não quem o escreveu: nenhum logo, nenhuma
cor de agência. O que dá identidade é a tipografia editorial (serifa, filete
duplo, versaletes) e a paleta do próprio produto, que vem no JSON. Assim o
documento parece ter saído de dentro do sistema que ele descreve.

Imprime pelo Chrome (Playwright), então o layout é 100% CSS e a paginação é a
do navegador — nada de Word.

FORMATO DO documento.json
-------------------------
{
  "produto": "Bússola Tributária",          // a manchete do cabeçalho
  "assinatura": "AFL Consultores",          // quem assina (topo e rodapé)
  "tipo": "Documento de acompanhamento",    // canto direito do cabeçalho
  "subtitulo": "Onde estamos e o que vem a seguir",
  "data": "4 de setembro de 2026",          // default: hoje, em português
  "abertura": "Parágrafo de abertura, com **negrito** onde importa.",
  "tokens": { "papel": "#f4f5f0", "tinta": "#141a17" },   // opcional, ver PADRAO
  "fontes": { "titulo": "Georgia, serif" },               // opcional
  "fontesArquivos": [                                     // opcional: .woff2 do produto
    { "familia": "Alegreya", "arquivo": "D:/.../alegreya.woff2", "peso": "700" }
  ],
  "partes": [
    { "titulo": "Hoje temos",
      "respiro": false,                     // true = mais ar antes do título
      "blocos": [ ... ],                    // opcional, antes dos itens
      "itens": [
        { "num": "01", "titulo": "A porta de entrada",
          "quebra": false,                  // true = começa em página nova
          "blocos": [ ... ] }
      ] }
  ],
  "rodape": "Uma linha sobre como o documento foi feito.",
  "saida": "D:/.../nome-do-documento.pdf"
}

BLOCOS (uma chave por bloco, na ordem em que você escreveu)
-----------------------------------------------------------
{"p": "parágrafo com **negrito**, _itálico_ e `mono`"}
{"bullets": ["item", "outro item"]}
{"figura": {"img": "01-login.png", "legenda": "O que olhar aqui.",
            "largura": 82,                  // % da coluna — é decisão de PAGINAÇÃO
            "retrato": false}}              // true = imagem em pé (PDF, celular)
{"boxe": {"rotulo": "Limitação de hoje", "tipo": "alerta", "texto": "..."}}
                                            // tipo: info (default) | alerta | ok
{"tabela": {"cabecalho": ["O que falta", "Situação"],
            "linhas": [["Abrir o link", "planejado"]],
            "estreita": true,               // não estica até a margem
            "numericas": [1]}}              // colunas alinhadas à direita
{"quebra": true}                            // quebra de página solta
"""

from __future__ import annotations

import base64
import json
import sys
from datetime import date
from pathlib import Path
from typing import Any

MESES = [
    "janeiro", "fevereiro", "março", "abril", "maio", "junho",
    "julho", "agosto", "setembro", "outubro", "novembro", "dezembro",
]

# Papel de ofício claro e tinta com viés verde: a direção editorial do modelo.
# Qualquer produto pode sobrescrever pelo "tokens" do JSON.
PADRAO = {
    "papel": "#f4f5f0",
    "realce": "#e7eae1",
    "tinta": "#141a17",
    "tinta2": "#454f49",
    "tinta3": "#6b746e",
    "filete": "#c5cbbe",
    "forte": "#141a17",
    "alerta": "#b45309",
    "ok": "#15803d",
}

FONTES_PADRAO = {
    "titulo": 'Georgia, "Times New Roman", serif',
    "corpo": 'Georgia, "Times New Roman", serif',
    "mono": 'Consolas, Menlo, monospace',
}


def esc(texto: Any) -> str:
    return (
        str("" if texto is None else texto)
        .replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
    )


def rico(texto: Any) -> str:
    """**negrito**, _itálico_ e `mono` no meio do texto."""
    import re

    saida = esc(texto)
    saida = re.sub(r"\*\*(.+?)\*\*", r"<strong>\1</strong>", saida)
    saida = re.sub(r"(?<!\w)_(.+?)_(?!\w)", r"<em>\1</em>", saida)
    saida = re.sub(r"`(.+?)`", r'<code class="mono">\1</code>', saida)
    return saida


def data_uri(caminho: Path) -> str | None:
    if not caminho.exists():
        print(f"  ! imagem não encontrada: {caminho}")
        return None
    ext = caminho.suffix.lstrip(".").lower()
    mime = "image/jpeg" if ext in ("jpg", "jpeg") else f"image/{ext}"
    return f"data:{mime};base64," + base64.b64encode(caminho.read_bytes()).decode()


def css(tokens: dict[str, str], fontes: dict[str, str], faces: str) -> str:
    vars_css = "".join(f"--{k}:{v};" for k, v in tokens.items())
    return f"""
{faces}
:root{{{vars_css}}}
@page{{size:A4}}
*{{box-sizing:border-box}}
body{{margin:0;background:var(--papel);color:var(--tinta);
     font-family:{fontes['corpo']};font-size:11px;line-height:1.5}}
.mono{{font-family:{fontes['mono']}}}
code.mono{{font-size:.92em}}

.masthead{{margin-bottom:6mm}}
.servico{{display:flex;justify-content:space-between;font-size:8px;
         letter-spacing:1.5px;text-transform:uppercase;color:var(--tinta3);padding-bottom:3px}}
h1.marca{{margin:0;font-family:{fontes['titulo']};font-size:40px;font-weight:700;
         letter-spacing:-.5px;line-height:1}}
.r1{{height:3px;background:var(--forte);margin-top:4mm}}
.r2{{height:1px;background:var(--forte);opacity:.55;margin-top:2px}}
.datalinha{{display:flex;justify-content:space-between;padding-top:4px;
           font-size:8.5px;letter-spacing:.6px;color:var(--tinta2)}}

.abertura{{font-size:12.5px;line-height:1.55;max-width:150mm;margin:0 0 6mm}}

h2.parte{{margin:8mm 0 4mm;font-family:{fontes['titulo']};font-size:13px;font-weight:700;
         letter-spacing:2px;text-transform:uppercase;border-bottom:2px solid var(--forte);
         padding-bottom:3px;page-break-after:avoid}}
h2.parte.respiro{{margin-top:14mm}}
h3.item{{margin:6mm 0 1.5mm;font-family:{fontes['titulo']};font-size:15px;font-weight:700;
        line-height:1.2;page-break-after:avoid}}
/* O título não pode ficar sozinho no pé da página, e o texto dele também não:
   a corrente título -> lide -> figura viaja junta. */
h3.item + p{{page-break-after:avoid}}
h3.item .num{{font-family:{fontes['mono']};font-size:11px;color:var(--tinta3);margin-right:6px}}
p{{margin:0 0 2.5mm;max-width:165mm;orphans:2;widows:2}}

/* A largura da imagem é decisão de PAGINAÇÃO, não de estética: a 100% da
   coluna (178mm) uma captura 16:9 fica com ~111mm de altura, o item passa de
   140mm e só cabe UM por página — sobra um quarto de folha em branco no pé.
   A 82% ela cai para ~91mm, o item fecha em ~121mm e dois itens dividem a
   página. Meça no PDF, não no navegador. */
figure{{margin:3mm 0 6mm;page-break-inside:avoid}}
figure img{{margin:0 auto;display:block;border:1px solid var(--filete)}}
.legenda{{font-size:9px;color:var(--tinta3);margin:1.5mm 0 0;font-style:italic;text-align:center}}

.boxe{{background:var(--realce);border-left:3px solid var(--forte);padding:3mm 4mm;
      margin:3mm 0 5mm;page-break-inside:avoid}}
.boxe .rot{{font-family:{fontes['mono']};font-size:8px;letter-spacing:1.4px;
           text-transform:uppercase;color:var(--tinta3);margin-bottom:1.5mm}}
.boxe.alerta{{border-left-color:var(--alerta)}}
.boxe.ok{{border-left-color:var(--ok)}}
.boxe p:last-child{{margin-bottom:0}}

table{{width:100%;border-collapse:collapse;margin:2mm 0 4mm;font-size:10px;
      page-break-inside:avoid}}
th{{text-align:left;font-size:8px;letter-spacing:1px;text-transform:uppercase;
   color:var(--tinta3);border-bottom:1px solid var(--forte);padding:1.5mm 2mm 1.2mm 0}}
td{{padding:1.5mm 2mm 1.5mm 0;border-bottom:.5pt solid var(--filete);vertical-align:top}}
td.num,th.num{{text-align:right;font-variant-numeric:tabular-nums;white-space:nowrap}}
table.estreita{{max-width:110mm}}

ul{{margin:0 0 3mm;padding-left:5mm}}
li{{margin-bottom:1.5mm}}
.quebra{{page-break-before:always}}
.rodape{{margin-top:8mm;border-top:1px solid var(--forte);padding-top:2mm;
        font-size:8.5px;color:var(--tinta3)}}
"""


def faces_css(fontes_arquivos: list[dict[str, str]]) -> str:
    """Embute as .woff2 do produto, se o JSON apontar para elas.

    Fonte que não existe nesta máquina é ignorada em silêncio: o documento cai
    na serifa de sistema e a paginação continua a mesma, em vez de quebrar.
    """
    partes = []
    for face in fontes_arquivos:
        caminho = Path(face["arquivo"])
        if not caminho.exists():
            print(f"  ! fonte não encontrada, usando fallback: {caminho}")
            continue
        b64 = base64.b64encode(caminho.read_bytes()).decode()
        partes.append(
            f"@font-face{{font-family:'{face['familia']}';"
            f"font-weight:{face.get('peso', '400')};font-style:normal;font-display:block;"
            f"src:url(data:font/woff2;base64,{b64}) format('woff2')}}"
        )
    return "\n".join(partes)


def bloco_html(bloco: dict[str, Any], imgdir: Path) -> str:
    if "p" in bloco:
        return f'<p>{rico(bloco["p"])}</p>'

    if "bullets" in bloco:
        itens = "".join(f"<li>{rico(b)}</li>" for b in bloco["bullets"])
        return f"<ul>{itens}</ul>"

    if "figura" in bloco:
        fig = bloco["figura"]
        uri = data_uri(imgdir / fig["img"] if not Path(fig["img"]).is_absolute() else Path(fig["img"]))
        if not uri:
            return ""
        # Retrato (um PDF, uma tela de celular) já é alto: a mesma largura da
        # paisagem estouraria a página inteira.
        largura = fig.get("largura", 68 if fig.get("retrato") else 82)
        legenda = (
            f'<figcaption class="legenda">{rico(fig["legenda"])}</figcaption>'
            if fig.get("legenda")
            else ""
        )
        return (
            f'<figure><img src="{uri}" alt="" style="width:{largura}%">{legenda}</figure>'
        )

    if "boxe" in bloco:
        boxe = bloco["boxe"]
        tipo = boxe.get("tipo", "info")
        rot = f'<div class="rot">{esc(boxe["rotulo"])}</div>' if boxe.get("rotulo") else ""
        corpo = "".join(f"<p>{rico(t)}</p>" for t in _lista(boxe.get("texto")))
        return f'<div class="boxe {tipo}">{rot}{corpo}</div>'

    if "tabela" in bloco:
        tab = bloco["tabela"]
        numericas = set(tab.get("numericas", []))
        # As classes saem montadas ANTES do f-string: aspas e barras dentro de
        # f-string só valem do Python 3.12 em diante, e esta skill roda no venv
        # de qualquer projeto.
        def cel(tag: str, i: int, conteudo: str) -> str:
            classe = ' class="num"' if i in numericas else ""
            return "<" + tag + classe + ">" + conteudo + "</" + tag + ">"

        classe_tab = ' class="estreita"' if tab.get("estreita") else ""
        cabecalho = ""
        if tab.get("cabecalho"):
            celulas = "".join(
                cel("th", i, esc(c)) for i, c in enumerate(tab["cabecalho"])
            )
            cabecalho = f"<thead><tr>{celulas}</tr></thead>"
        corpo = "".join(
            "<tr>" + "".join(cel("td", i, rico(c)) for i, c in enumerate(linha)) + "</tr>"
            for linha in tab.get("linhas", [])
        )
        return f"<table{classe_tab}>{cabecalho}<tbody>{corpo}</tbody></table>"

    if "quebra" in bloco:
        return '<div class="quebra"></div>'

    raise ValueError(f"bloco desconhecido: {bloco}")


def _lista(valor: Any) -> list[Any]:
    if valor is None:
        return []
    return valor if isinstance(valor, list) else [valor]


def montar_html(spec: dict[str, Any], imgdir: Path) -> str:
    tokens = {**PADRAO, **spec.get("tokens", {})}
    fontes = {**FONTES_PADRAO, **spec.get("fontes", {})}
    faces = faces_css(spec.get("fontesArquivos", []))

    hoje = date.today()
    data = spec.get("data") or f"{hoje.day} de {MESES[hoje.month - 1]} de {hoje.year}"

    corpo = []
    for parte in spec.get("partes", []):
        respiro = " respiro" if parte.get("respiro") else ""
        if parte.get("quebra"):
            corpo.append('<div class="quebra"></div>')
        if parte.get("titulo"):
            corpo.append(f'<h2 class="parte{respiro}">{esc(parte["titulo"])}</h2>')
        for bloco in parte.get("blocos", []):
            corpo.append(bloco_html(bloco, imgdir))
        for item in parte.get("itens", []):
            if item.get("quebra"):
                corpo.append('<div class="quebra"></div>')
            num = f'<span class="num">{esc(item["num"])}</span>' if item.get("num") else ""
            corpo.append(f'<h3 class="item">{num}{esc(item["titulo"])}</h3>')
            for bloco in item.get("blocos", []):
                corpo.append(bloco_html(bloco, imgdir))

    rodape = (
        f'<div class="rodape">{rico(spec["rodape"])}</div>' if spec.get("rodape") else ""
    )
    abertura = (
        f'<p class="abertura">{rico(spec["abertura"])}</p>' if spec.get("abertura") else ""
    )

    return f"""<!doctype html><html lang="pt-BR"><head><meta charset="utf-8">
<title>{esc(spec['produto'])} — {esc(spec.get('subtitulo', ''))}</title>
<style>{css(tokens, fontes, faces)}</style></head><body>

<header class="masthead">
  <div class="servico"><span>{esc(spec.get('assinatura', ''))}</span>
    <span>{esc(spec.get('tipo', 'Documento de acompanhamento'))}</span></div>
  <h1 class="marca">{esc(spec['produto'])}</h1>
  <div class="r1"></div><div class="r2"></div>
  <div class="datalinha"><span>{esc(spec.get('subtitulo', ''))}</span>
    <span class="mono">{esc(data)}</span></div>
</header>

{abertura}
{''.join(corpo)}
{rodape}

</body></html>"""


def main(argv: list[str]) -> int:
    if len(argv) < 2:
        print("uso: gerar_pdf.py <documento.json> [pasta-das-imagens]")
        return 2

    spec_file = Path(argv[1])
    spec = json.loads(spec_file.read_text(encoding="utf-8"))
    imgdir = Path(argv[2]) if len(argv) > 2 else spec_file.resolve().parent

    html = montar_html(spec, imgdir)

    destino = Path(spec["saida"])
    destino.parent.mkdir(parents=True, exist_ok=True)

    # O HTML fica junto do JSON (scratchpad), não junto do PDF: só o PDF vai
    # para a pasta do projeto. Ajuda a depurar layout no Chrome de verdade —
    # mas confira a PAGINAÇÃO no PDF, porque o navegador não mostra onde a
    # folha corta.
    lado = spec_file.resolve().with_suffix(".html")
    lado.write_text(html, encoding="utf-8")

    from playwright.sync_api import sync_playwright

    rodape_pag = (
        f'<div style="width:100%;font-family:{spec.get("fontes", {}).get("corpo", "Georgia, serif")};'
        'font-size:8px;color:#6b746e;padding:0 16mm;display:flex;'
        'justify-content:space-between">'
        f'<span>{esc(spec["produto"])}'
        + (f' · {esc(spec["assinatura"])}' if spec.get("assinatura") else "")
        + '</span><span class="pageNumber"></span></div>'
    )

    with sync_playwright() as p:
        nav = p.chromium.launch()
        pg = nav.new_page()
        pg.set_content(html, wait_until="load")
        pg.wait_for_timeout(700)
        pg.pdf(
            path=str(destino),
            format="A4",
            print_background=True,
            margin={"top": "16mm", "bottom": "14mm", "left": "16mm", "right": "16mm"},
            display_header_footer=True,
            header_template="<div></div>",
            footer_template=rodape_pag,
        )
        nav.close()

    kb = destino.stat().st_size // 1024
    print(f"PDF salvo em: {destino}  ({kb} KB)")
    print(f"HTML de depuração: {lado}")
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))
